dbms/src/Flash/Executor/DataStreamExecutor.cpp

```diff
@@ -65,11 +69,23 @@ String DataStreamExecutor::toString() const
 
 int DataStreamExecutor::estimateNewThreadCount()
 {
-    return data_stream->estimateNewThreadCount();
+    return estimate_thread_cnt;
 }
 
 RU DataStreamExecutor::collectRequestUnit()
 {
-    return toRU(data_stream->estimateCPUTimeNs());
+    auto origin_cpu_time_ns = data_stream->estimateCPUTimeNs();
+    UInt64 total_thread_cnt = thread_cnt_before_execute + estimate_thread_cnt;
+    size_t physical_cpu_cores = getNumberOfPhysicalCPUCores();
+    if (origin_cpu_time_ns <= 0 || total_thread_cnt <= physical_cpu_cores)
+        return toRU(origin_cpu_time_ns);
+
+    // When the number of threads is greater than the number of cpu cores,
+    // BlockInputStream's estimated cpu time will be much greater than the actual value.
+    double per_thread_cpu_time_ns = static_cast<double>(origin_cpu_time_ns) / estimate_thread_cnt;
```

Comment:
why the final result is origin_cpu_time_ns / estimate_thread_cnt/ (estimate_thread_cnt+K) * physical_cpu_cores * physical_cpu_cores ?

Reply:
Since a cpu core executes several threads, think of `execute_time_of_thread_A = cputime_of_thread_A + cputime_of_other_thread`.
Assuming that the cputime of other threads is the same as thread A, then we have `execute_time_of_thread_A = total_thread_cnt * cputime_of_thread_A`.
Assuming that the number of threads is divided equally among all cpu cores, then `the number of threads allocated to cpu core A = total_thread_cnt / cpu core num`.

Reply:
But there is still no way to eliminate the effect of `condition_cv.wait` here...

Reply:
I've added some derivations to the comments.
Now the final result is `ceil(static_cast<double>(execute_time_ns) / total_thread_cnt * logical_cpu_cores)`